### DISCUSSION TITLE
Looking to host Labelme on a server

### DISCUSSION THREAD

**Discussion by Prady96** [wkentaro/labelme#1405] [Ideas / Feature requests]
We have multiple users using labelme parallelly, can we deploy Labelme on the server?

In a way multiple users can run the same instance together in one machine.

This would help us to save on the computation cost of our system.

**Comment by wkentaro** [comment]
@Prady96 Interesting. Can I ask a couple questions to understand your requirements?

About the device:

1) How currently users access Labelme? On a powerful laptop machine?
2) How would the user access to Labelme after deployment? From a mobile phone / cheap laptop machine via browser/vnc-viewer?

About what to share between users:

3) Do you want to let users to share a mouse to annotate a single image?
4) Do you want to let users to share unlabeled images to annotate different images in parallel?

About the functionality:

5) Do the users use Labelme's advanced functionality of AI-annotation (ai-mask, ai-polygon)? Or basic functionality only (e.g., polygon, box annotations).
6) If you're using only basic functionality, still that's computationally heavy to run on a cheap machine?

**Reply by Prady96** [reply]
Hi @wkentaro 

**1. How do current users access Labelme? On a powerful laptop machine?**
At present, we've implemented n-computing wherein five users utilize a single Xeon server. Each user operates independently, resulting in five instances of Python Anaconda running simultaneously on the same machine.

**2. How would the user access to Labelme after deployment?** 
We can operate it on our server and provide an IP for them to label images on their devices. If we scale this system for 15 users, we won't require 15 separate Python Anaconda environments. We want a single LabelMe operating on a powerful machine, capable of handling 15 users simultaneously.

About what to share between users:

**3. Do you want to let users to share a mouse to annotate a single image?**
Implementing this could potentially decrease our quality control time per image. I've observed this feature in higher-end products, but their cost is too expensive for our startup.

**4. Do you want to let users to share unlabeled images to annotate different images in parallel?**
I would love to see that feature.

About the functionality:

**5. Do the users use Labelme's advanced functionality of AI-annotation (ai-mask, ai-polygon)? Or basic functionality only (e.g., polygon, box annotations).**
Indeed, we employ it internally on every machine. It does take nearly 2 minutes to load after startup, but proves to be a time-saver during labelling. If we can optimise the usage of labelme, we hold the potential to scale this system.

**6. If you're using only basic functionality, still that's computationally heavy to run on a cheap machine?**
The primary issue I've encountered is that the current setup doesn't allow for hosting similar to a website. Ideally, we should be able to host on a VM on a high-performance machine where a single instance of LabelMe can be shared among all users. However, the current situation creates a new instance of LabelMe for each user, resulting in an inefficient use of resources.

Any resolution to this issue would be greatly appreciated.

**Reply by wkentaro** [reply]
@Prady96 Hi. I have further questions.

> At present, we've implemented n-computing wherein five users utilize a single Xeon server. Each user operates independently, resulting in five instances of Python Anaconda running simultaneously on the same machine.

Do you currently connect 5 monitors and keyboards so that the 5 users can simultaneously use Labelme?

I need to understand what computational resource you want to reduce. Even if you host Labelme on a Website, the users would need computers to access it, so the cost of devices doesn't change.

> We can operate it on our server and provide an IP for them to label images on their devices. If we scale this system for 15 users, we won't require 15 separate Python Anaconda environments. We want a single LabelMe operating on a powerful machine, capable of handling 15 users simultaneously.

What prevents the users install Python and Labelme on their local devices?

> Indeed, we employ it internally on every machine. It does take nearly 2 minutes to load after startup, but proves to be a time-saver during labelling. If we can optimise the usage of labelme, we hold the potential to scale this system.

Do you say the AI-annotation is computationally heavy and require server? Or even the basic functionality requires server? (I don't think basic functionality should use too much memory and cpu at the level of requiring high-end server)

I need to understand why you need high-end server instead of a local computer for each user.

**Reply by Prady96** [reply]
Thank you @wkentaro, for taking out time.

> Do you currently connect 5 monitors and keyboards so that the 5 users can simultaneously use Labelme?

> I need to understand what computational resource you want to reduce. Even if you host Labelme on a Website, the users would need computers to access it, so the cost of devices doesn't change.

Please try to understand the point.

If today you want to host a website then you run 1 Apache/nginx server and vertically scale the system to accommodate the influx of users. However, the current system is operating a unique nginx server for each individual user. For instance, five users equate to five nginx servers.

The crux of the issue lies here: there should be one anaconda python running on the server that can be shared by all five users. As it stands now, when I open the task manager, five python instances are shown running on the admin.

> What prevents the users install Python and Labelme on their local devices?

Moreover, it's important to note that annotations are often performed by non-tech workers, and sometimes part-time workers, who may not be proficient with command line. Once distributed using the IP, they can easily work within the browser. Should any issue arise, we could handle it server-side.

> Do you say the AI-annotation is computationally heavy and require server? Or even the basic functionality requires server? (I don't think basic functionality should use too much memory and cpu at the level of requiring high-end server)

Currently, this shouldn't be regarded as a major concern. We can address it through vertical scaling, with the ideal set-up being one Anaconda Python server catering to all users simultaneously, as opposed to the current situation. Each machine maintains its dedicated resources.

**Reply by wkentaro** [reply]
Okay.

If you really want to host Labelme on browser, it is just not possible as Labelme is Python+Qt application. You need to adopt different software like [LabelMeAnnotationTool](https://github.com/CSAILVision/LabelMeAnnotationTool).

To run multiple users asynchronously without too much load on the server, we need to run the program on the users' computer instead of the server. This is just not possible with Python on the browsers, and that's why we need to use Javascript-based web app for that. (because Javascript runs on a browser so doesn't need much computation on the server)

**Reply by Prady96** [reply]
We've tested this tool. It's operational, but lacks AI Annotation. We can accommodate a high server load, yet it should function browser-based, instead of requiring installation on individual devices.

**Reply by Prady96** [reply]
Is there any simple way to deploy it on the browser?

**Reply by wkentaro** [reply]
Only possibility I can think of is:

- Launch Labelme with multiple python instances
- Access each via VNC viewer (probably possible via a browser)

But you can't avoid launching multiple python.

**Reply by Prady96** [reply]
Both of them we are currently doing taking too much resources. How other tools are working?

**Reply by wkentaro** [reply]
Can you give me the examples for "other tools"? If you're talking about browser-based, they are using Javascript, so most of the operation runs on the user side instead of server side.

**Reply by Prady96** [reply]
Like CVAT, RoboFlow or V7?

We have the necessary infrastructure, but we're in need of a software solution that can be installed and operated on our own servers to maintain data privacy.

**Reply by wkentaro** [reply]
CVAT is Javascript-based.

<img src=https://github.com/labelmeai/labelme/assets/4310419/8487ae72-67f2-4932-999a-8cb937bbe329 width=500>
